sdk/core/Azure.Core/src/Shared/OperationPoller.cs

```diff
@@ -6,96 +6,99 @@
             }
         }
 
-        public virtual ValueTask<Response<T>> WaitForCompletionAsync<T>(Operation<T> operation, TimeSpan? suggestedInterval, CancellationToken cancellationToken) where T : notnull
-           => WaitForCompletionAsync(operation.UpdateStatusAsync, () => operation.HasCompleted, () => operation.Value, operation.GetRawResponse, suggestedInterval, cancellationToken);
-
-        public virtual async ValueTask<Response<T>> WaitForCompletionAsync<T>(UpdateStatusAsync updateStatusAsync, HasCompleted hasCompleted, Value<T> value, GetRawResponse getRawResponse, TimeSpan? suggestedInterval, CancellationToken cancellationToken)
+        private async ValueTask<Response> WaitForCompletionAsync(bool async, OperationInternalBase operation, TimeSpan? suggestedInterval, CancellationToken cancellationToken)
         {
             while (true)
             {
-                Response response = await updateStatusAsync(cancellationToken).ConfigureAwait(false);
-
-                if (hasCompleted())
+                Response response = async ? await operation.UpdateStatusAsync(cancellationToken).ConfigureAwait(false) : operation.UpdateStatus(cancellationToken);
+                if (operation.HasCompleted)
                 {
-                    return Response.FromValue(value(), getRawResponse());
+                    return operation.RawResponse;
                 }
 
-                await Task.Delay(_delayStrategy.GetNextDelay(response, suggestedInterval), cancellationToken).ConfigureAwait(false);
+                await Delay(async, _delayStrategy.GetNextDelay(response, suggestedInterval), cancellationToken).ConfigureAwait(false);
             }
         }
 
-        public virtual Response<T> WaitForCompletion<T>(Operation<T> operation, TimeSpan? suggestedInterval, CancellationToken cancellationToken) where T : notnull
-           => WaitForCompletion(operation.UpdateStatus, () => operation.HasCompleted, () => operation.Value, operation.GetRawResponse, suggestedInterval, cancellationToken);
-
-        public virtual Response<T> WaitForCompletion<T>(UpdateStatus updateStatus, HasCompleted hasCompleted, Value<T> value, GetRawResponse getRawResponse, TimeSpan? suggestedInterval, CancellationToken cancellationToken)
+        private static async ValueTask Delay(bool async, TimeSpan delay, CancellationToken cancellationToken)
         {
-            while (true)
+            if (async)
             {
-                Response response = updateStatus(cancellationToken);
-
-                if (hasCompleted())
-                {
-                    return Response.FromValue(value(), getRawResponse());
-                }
-
-                Thread.Sleep(_delayStrategy.GetNextDelay(response, suggestedInterval));
+                await Task.Delay(delay, cancellationToken).ConfigureAwait(false);
+            }
+            else if (cancellationToken.CanBeCanceled)
+            {
+                cancellationToken.WaitHandle.WaitOne(delay);
```

Comment:
Should we throw if the token is canceled here?

Reply:
Awesome catch!